src/main/java/apoc/cypher/Cypher.java

```diff
@@ -95,26 +112,47 @@
         boolean addStatistics = Util.toBoolean(config.getOrDefault("statistics",true));
         int timeout = Util.toInteger(config.getOrDefault("timeout",10));
+        int queueCapacity = Util.toInteger(config.getOrDefault("queueCapacity",100));
         List<RowResult> result = new ArrayList<>();
         for (String f : fileNames) {
-            List<RowResult> rowResults = runManyStatements(readerForFile(f), Collections.emptyMap(), true, addStatistics, timeout).collect(Collectors.toList());
+            List<RowResult> rowResults = runManyStatements(readerForFile(f), Collections.emptyMap(), true, addStatistics, timeout, queueCapacity).collect(Collectors.toList());
             result.addAll(rowResults);
         }
         return result.stream();
     }
 
-    private Stream<RowResult> runManyStatements(Reader reader, Map<String, Object> params, boolean schemaOperation, boolean addStatistics, int timeout) {
-        BlockingQueue<RowResult> queue = new ArrayBlockingQueue<>(100);
-        Util.inThread(() -> {
+    private Stream<RowResult> runManyStatements(Reader reader, Map<String, Object> params, boolean schemaOperation, boolean addStatistics, int timeout, int queueCapacity) {
+        BlockingQueue<RowResult> queue = new ArrayBlockingQueue<>(queueCapacity);
+        runInSeparateThreadAndSendTombstone(() -> {
             if (schemaOperation) {
-                runSchemaStatementsInTx(reader, queue, params, addStatistics,timeout);
+                runSchemaStatementsInTx(reader, queue, params, addStatistics, timeout);
             } else {
-                runDataStatementsInTx(reader, queue, params, addStatistics,timeout);
+                runDataStatementsInTx(reader, queue, params, addStatistics, timeout);
             }
-            queue.put(RowResult.TOMBSTONE);
-            return null;
-        });
-        return StreamSupport.stream(new QueueBasedSpliterator<>(queue, RowResult.TOMBSTONE, terminationGuard, timeout), false);
+        }, queue, RowResult.TOMBSTONE);
+        return StreamSupport.stream(new QueueBasedSpliterator<>(queue, RowResult.TOMBSTONE, terminationGuard), false);
+    }
+
+
+    private <T> void runInSeparateThreadAndSendTombstone(Runnable action, BlockingQueue<T> queue, T tombstone) {
+        /* NB: this must not be called via an existing thread pool - otherwise we could run into a deadlock
+           other jobs using the same pool might completely exhaust at and the thread sending TOMBSTONE will
+           wait in the pool's job queue.
+         */
+        new Thread(() -> {
+            try {
+                action.run();
+            } finally {
+                while (true) {  // ensure we send TOMBSTONE even if there's an InterruptedException
+                    try {
+                        queue.put(tombstone);
+                        return;
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
```

Comment:
can this lead into being blocked forever? Shouldn't we take the timeout into account?
int put and/or the whole method time?

Reply:
Blocking forever?